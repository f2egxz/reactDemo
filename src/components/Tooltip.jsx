//-*- mode: react -*-
//-*- coding: utf-8 -*-

/**
 * Tooltip.jsx
 *
 * -TODO 文字提示框(出现位置位于文字：上左，上中，上右，右侧中间，下右，下中，下左，左侧中间)
 *
 * @author fe.gxz
 */

import React          from 'react';
import { classnames } from 'util/view-extra'
import Styles         from 'style/Tooltip.css'

function boundClassNameByType (placement){
  switch (placement) {
    case 'Top':
      return classnames(Styles.tipTC);
    case 'Right':
      return classnames(Styles.tipRC);
    case 'Bottom':
      return classnames(Styles.tipBC);
    case 'Left':
      return classnames(Styles.tipLC);
    case 'TL':
      return classnames(Styles.tipTL);
    case 'TR':
      return classnames(Styles.tipTR);
    case 'BL':
      return classnames(Styles.tipBL);
    case 'BR':
      return classnames(Styles.tipBR);
  }
}

/**
 * Tooltip
 *
 * 文本提示框组件
 *
 * @param {String} text
 * @param {String} placement
 *
 * @return {html} Tooltip
 */
class Tooltip extends React.Component {
  constructor(){
    super()
    this.state = {
      show: false
    }
  }
  render(){
    const {placement,children,text} = this.props
    const tip = <span className={boundClassNameByType(placement)}>{text}</span>
    return (
      <div className = {classnames(Styles.tooltipBox)}
           onMouseEnter = {this.handleMouseEnter.bind(this)}
           onMouseLeave={this.handleMouseLeave.bind(this)}>
        { children }
        {this.state.show?tip:''}
      </div>
    )
  }
  handleMouseEnter(){
    this.setState({
      show:true
    })
  }
  handleMouseLeave(){
    this.setState({
      show:false
    })
  }
}

Tooltip.propTypes = {
  text:React.PropTypes.string.isRequired,
  placement:React.PropTypes.oneOf([
    'Top',
    'Right',
    'Bottom',
    'Left',
    'TL',
    'TR',
    'BL',
    'BR'
  ])
}

Tooltip.defaultProps = {
  placement: 'Top'
}



export default Tooltip
