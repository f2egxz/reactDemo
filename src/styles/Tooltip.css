/*-*- mode: css -*-*/
/*-*- coding: utf-8 -*-*/

/**
 * Tooltip.css
 *
 * 文本提示框
 *
 * @author fe.gxz
 */
@value ThemeBgc:  #6e6e6e;


/**
 * Tooltip
 *  文本提示框的样式
 *  上左，上中，上右
 *  右中， 左中
 *  下左，下中，下右
 */


.tipTL {
  left: 0;
  top: 0;
  transform: translate(0, calc(-100% - 10px));
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;

&::after {
   content: '';
   position: absolute;
   bottom: 0;
   border-top: 6px solid ThemeBgc;
   border-left: 6px solid transparent;
   border-right: 6px solid transparent;
   left: 0;
   transform: translate(100%,100%);
   }
}

.tipTC{
  left:50%;
  top:0;
  transform: translate(-50%,calc(-100% - 10px));
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   bottom: 0;
   border-top: 6px solid ThemeBgc;
   border-left: 6px solid transparent;
   border-right: 6px solid transparent;
   left: 50%;
   transform: translate(-50%,100%);
 }
}

.tipTR{
  top:0;
  right: 0;
  transform: translate(0,calc(-100% - 10px));
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   bottom: 0;
   border-top: 6px solid ThemeBgc;
   border-left: 6px solid transparent;
   border-right: 6px solid transparent;
   right: 0;
   transform: translate(-100%,100%);
 }
}


.tipRC{
  top:50%;
  right: 0;
  transform: translate(calc(100% + 10px),-50%);
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   left: 0;
   border-right: 6px solid ThemeBgc;
   border-top: 6px solid transparent;
   border-bottom: 6px solid transparent;
   top: 50%;
   transform: translate(-100%,-50%);
 }
}


.tipLC{
  top: 50%;
  left: 0;
  transform: translate(calc(-100% - 10px),-50%);
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   right: 0;
   border-left: 6px solid ThemeBgc;
   border-top: 6px solid transparent;
   border-bottom: 6px solid transparent;
   top: 50%;
   transform: translate(100%,-50%);
 }
}



.tipBL  {
  bottom: 0;
  left: 0;
  transform: translate(0,calc(100% + 10px));
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   top: 0;
   border-bottom: 6px solid ThemeBgc;
   border-left: 6px solid transparent;
   border-right: 6px solid transparent;
   left: 0;
   transform: translate(100%,-100%);
 }
}

.tipBC{
  bottom: 0;
  left: 50%;
  transform: translate(-50%,calc(100% + 10px));
  /*composes: tipBox;*/
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   top: 0;
   border-bottom: 6px solid ThemeBgc;
   border-left: 6px solid transparent;
   border-right: 6px solid transparent;
   left: 50%;
   transform: translate(-50%,-100%);
 }
}

.tipBR{
  bottom: 0;
  right: 0;
  transform: translate(0,calc(100% + 10px));
  position: absolute;
  background-color: ThemeBgc;
  padding: 5px;
  color:#fff;
  border-radius: 4px;
&::after {
   content: '';
   position: absolute;
   top: 0;
   border-bottom: 6px solid ThemeBgc;
   border-left: 6px solid transparent;
   border-right: 6px solid transparent;
   right: 0;
   transform: translate(-100%,-100%);
 }
}


/**
 * tooltipBox 的样式
 */
.tooltipBox {
  position: relative;
  cursor: default;
}
