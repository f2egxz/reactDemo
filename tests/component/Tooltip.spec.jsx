import React from 'react';
import Tooltip from 'component/Tooltip.jsx';
import renderer from 'react-test-renderer';

test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <Tooltip text="test">this is test content</Tooltip>
  );
  let Tooltips = component.toJSON();
  expect(Tooltips).toMatchSnapshot();

  // manually trigger the callback
  Tooltips.props.onMouseEnter();
  // re-rendering
  Tooltips = component.toJSON();
  expect(Tooltips).toMatchSnapshot();

  // manually trigger the callback
  Tooltips.props.onMouseLeave();
  // re-rendering
  Tooltips = component.toJSON();
  expect(Tooltips).toMatchSnapshot();
});
