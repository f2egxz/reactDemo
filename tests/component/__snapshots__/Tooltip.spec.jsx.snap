exports[`test Link changes the class when hovered 1`] = `
<div
  className="tooltipBox"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  this is test content
  
</div>
`;

exports[`test Link changes the class when hovered 2`] = `
<div
  className="tooltipBox"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  this is test content
  <span
    className="tipTC">
    test
  </span>
</div>
`;

exports[`test Link changes the class when hovered 3`] = `
<div
  className="tooltipBox"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  this is test content
  
</div>
`;
